Kernel/TTY/MasterPTY.cpp

```diff
@@ -96,7 +96,7 @@ bool MasterPTY::can_write_from_slave() const
 {
     if (m_closed)
         return true;
-    return m_buffer->space_for_writing();
+    return m_buffer->space_for_writing() >= 2;
```

Comment:
Is this the correct fix for the problem? the problem is that we translate "\n" to "\r\n", so we write 2 chars instead of one, but the syscall handler can't understand why this happens. What other OSes do about this behavior?

Reply:
Linux's TTY write [function](https://elixir.bootlin.com/linux/v6.5-rc1/source/drivers/tty/n_tty.c#L2334) does something similar where it tries to print characters out in a while-loop. If the buffer does not have enough space (for eg [here](https://elixir.bootlin.com/linux/v6.5-rc1/source/drivers/tty/n_tty.c#L436)), they retry writing in the next iteration of the while-loop.

Serenity's write syscall handler calls `can_write(...)` on files to check if there is space to write at least 1 byte and hence expects this [check](https://github.com/amansingh-13/serenity/blob/d823e0b82324779396a43a0eff76be31723119e7/Kernel/Syscalls/write.cpp#L95) to succeed. However, in TTYs, we may need 2 bytes empty to write out at least 1 byte from the input string. Instead of ensuring we always have at least 2 bytes empty, the check could be made dependent on the character we are writing (2 bytes for "\n" and 1 for rest), but that would require changing much more code.

Reply:
The question is probably if this is correct to block a user on this, or rather we should immediately return with an error like `ENOBUFS`...

Reply:
That probably depends on the `O_NONBLOCK` option. I think this is [handled](https://github.com/amansingh-13/serenity/blob/d823e0b82324779396a43a0eff76be31723119e7/Kernel/Syscalls/write.cpp#L71) by Serenity's write syscall handler?

Reply:
@supercomputer7 😶‍🌫️